newChat lib API

spolu opened the pull request:
This PR introduces a `newChat` lib function that will start a chat from a user query (create the session, trigger retrieval if needed, store messages, etc...) generating events that will let us build the Slack bot and eventually even back the chat product

Also push towards using Authenticator instead of owner: WorkspaceType when relevant.

spolu commented:
@philipperolet FYI it's a bit more involved than I expected but it shaping nicely

spolu commented:
Feel free to comment if you see anything weird, obviously this is still WIP.

philipperolet commented:
linting?

spolu commented:
@philipperolet should be good :+1:

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+      yield* handleError(
+        `Chat error: [${res.error.type}] ${res.error.message}`
+      );
+      break;
```
at that stage the error message will be stored in DB (+ change title etc), cf the discussion we had IRL on the topic
This changes the existing logic; we can do it but warrants a new discussion & clarification in the code IMHO

Other option is to keep the existing logic by throwing. The caller of newChat can then decide how to handle the error. Or just exit the generator with return

spolu replied:
Let's simply not store error message for now :+1: 

We can't throw, we're streaming so the API endpoint won't have an opportuinity to retrun an error

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+      }
+
+      if (event.type === "error") {
+        yield* handleError(event.content.message);
```
same as above comment wrt storing error message in session (as well as other calls to handleError)

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+
+          if (m.role === "assistant") {
+            if (!assistantMessageTriggered) {
+              assistantMessageTriggered = true;
```
optional: would suggest cutting in a few functions, since it's getting a bit harder to read with the 6 levels of control blocks. I could take care of it in a following PR

spolu replied:
Yes they will have to be generators functions as we're in an async generator here. The handleError style can be used here indeed

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+
+  // Master loop that will run until the last message is not a "retrieval" request. It will call the
+  // assistant and push to the `messages` array. The assistant either push an "assistant" message
+  // with a response or a "retrieval" message that will contain a query, if that's the case we run
```
In front, at the time 2 retrievals are stored, one with the query & one with the docs
This was done with the idea that things happen in 2 steps, one with computing the params (query & later time, data sources, etc) and one with actually getting the docs , honestly more than default than conviction

We could switch to a one message logic, no strong convictions on this but would require a migration. We all love migrations :heart_decoration:

spolu replied:
I think it's a better model to have one message eventually :+1: 

I'll be happy to take that migration once we move the client to this logic :+1:

spolu replied:
Ah I get it this means we need a migration + change the client-side logic to map to this new logic because the conversations that will be created by this code will not be parseable by front

spolu replied:
Will the current client code be able to display this format? I guess yes no? In whic case we can do the migration later?

philipperolet reviewed (changes requested):
Two general points:
1. I think changing the fn to also handle ongoing chats is very straightforward (messages instead of userMessage + sessionId set or not), on another PR to pave the way for UI to be switched to this -- I could do it
2. Also interested by separating files, 1 for crud ops 1 for chat (more readable /findable, less scrolling) -- I could also do it :)

spolu commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+    DustProdActionRegistry["chat-assistant-wfn"].config
+  );
+
+  const assistantContext = {
```
Will leave that to you :)

spolu commented on front/lib/api/chat.ts:
```diff
@@ -42,9 +56,14 @@ export async function getChatSessions(
 ): Promise<ChatSessionType | null> {
+  const owner = auth.workspace();
+  if (!owner) {
+    return null;
```
As per IRL we may want to use `Result` in the future but not throwing for sure

spolu commented:
@philipperolet r? PTAL 
- addressed nits
- handled the fact that we don't store interactions that led to an error. The logic is pretty simple here as we have 1 interactions so we just don't store if we got an error (this will be the same for the follow-up implementation)

The main question that remains is storing just one retrieval message.

Is the UI today capable of displaying this format aka messages without retrieval message with query but no retrievals?

If so we don't need a migration just yet, only once we make the logic converge

spolu commented:
Also added throws for conditions that sould never happen. As this will not be caught in the API routes and will lead to a 500 and that's kind of what we want.

Basically it's fine to throw for a non-recoverable event

spolu commented:
Pushed some fixes but there are still a few bugs
I'm testing end to end right now
Will iron them out and merge :+1:

spolu commented:
Fully tested with
```
import { createParser } from "eventsource-parser";

import { Authenticator, getOrCreateSystemApiKey } from "@app/lib/auth";

const main = async () => {
  const auth = await Authenticator.internalBuilderForWorkspace("b809011d38");
  const owner = auth.workspace();
  if (!owner) {
    throw new Error("Invalid workspace");
  }
  const keyRes = await getOrCreateSystemApiKey(owner);
  if (keyRes.isErr()) {
    throw new Error(`Failed to get system API key: ${keyRes.error}`);
  }
  const key = keyRes.value;

  const res = await fetch(`http://localhost:3000/api/v1/w/b809011d38/chats`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${key.secret}`,
    },
    body: JSON.stringify({
      user_message: "Who are the design partners of Dust?",
      time_zone: "Europe/Paris",
    }),
  });

  if (!res.ok || !res.body) {
    throw new Error(`Failed to create chat`);
  }

  const parser = createParser((event) => {
    console.log(event);
  });

  const reader = res.body.getReader();

  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }
    parser.feed(new TextDecoder().decode(value));
  }
};

main()
  .then(() => {
    console.error("\x1b[32m%s\x1b[0m", `Done`);
    process.exit(0);
  })
  .catch((err) => {
    console.error("\x1b[31m%s\x1b[0m", `Error: ${err.message}`);
    console.log(err);
    process.exit(1);
  });
```

spolu commented:
And confirmed these can be accessed by link with the resulting messages format

spolu commented:
Added a max of 4 iterations.
Merging!

spolu commented:
@philipperolet pushed it to the finish line :grimacing: let me know what you think. There might still be a few things to iron out but it does work and the resulting chats are accessible by the UI :+1:

philipperolet commented:
Cool :fire: Was working on the fixes on my end but you beat me to it :) :pray: :pray:  for taking care of it, I'll have a look first thing monday

spolu commented:
Couldn’t resist the urge to finish it at the airport 😬 I think we were already flying when I merged this one 🙈

philipperolet commented:
:smile: #tmtc

philipperolet commented:
> Fully tested with
> 
> ```
> import { createParser } from "eventsource-parser";
> 
> import { Authenticator, getOrCreateSystemApiKey } from "@app/lib/auth";
> 
> const main = async () => {
>   const auth = await Authenticator.internalBuilderForWorkspace("b809011d38");
>   const owner = auth.workspace();
>   if (!owner) {
>     throw new Error("Invalid workspace");
>   }
>   const keyRes = await getOrCreateSystemApiKey(owner);
>   if (keyRes.isErr()) {
>     throw new Error(`Failed to get system API key: ${keyRes.error}`);
>   }
>   const key = keyRes.value;
> 
>   const res = await fetch(`http://localhost:3000/api/v1/w/b809011d38/chats`, {
>     method: "POST",
>     headers: {
>       "Content-Type": "application/json",
>       Authorization: `Bearer ${key.secret}`,
>     },
>     body: JSON.stringify({
>       user_message: "Who are the design partners of Dust?",
>       time_zone: "Europe/Paris",
>     }),
>   });
> 
>   if (!res.ok || !res.body) {
>     throw new Error(`Failed to create chat`);
>   }
> 
>   const parser = createParser((event) => {
>     console.log(event);
>   });
> 
>   const reader = res.body.getReader();
> 
>   while (true) {
>     const { done, value } = await reader.read();
>     if (done) {
>       break;
>     }
>     parser.feed(new TextDecoder().decode(value));
>   }
> };
> 
> main()
>   .then(() => {
>     console.error("\x1b[32m%s\x1b[0m", `Done`);
>     process.exit(0);
>   })
>   .catch((err) => {
>     console.error("\x1b[31m%s\x1b[0m", `Error: ${err.message}`);
>     console.log(err);
>     process.exit(1);
>   });
> ```

cc @lasryaric the code of stan to call the api, useful for you if you work for the slackbot I think

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -425,11 +425,8 @@ export async function* newChat(
   }
 
-  // create a Session
+  // Create a Session Id but do not store the session just now.
```
move this line to where we actually store?

philipperolet commented on front/lib/api/chat.ts:
```diff
@@ -299,3 +325,375 @@ export async function updateChatMessageFeedback({
+                role: "assistant",
+              } as ChatMessageTriggerEvent;
+            }
+            yield { message: m } as ChatMessageCreateEvent;
```
missing type here -- NOOP since added in chat eval PR that will be merged today

philipperolet reviewed (commented):
LGTM! 2 minor coms, 1 already adressed, the ohter optional
